src/react/SessionManager.ts

```diff
@@ -0,0 +1,267 @@
+/*
+   This file is part of Astarte.
+
+   Copyright 2020 Ispirata Srl
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+import { createNanoEvents, Emitter, Unsubscribe } from 'nanoevents';
+
+type Storage = {
+  getItem: (key: string) => string | null;
+  removeItem: (key: string) => void;
+  setItem: (key: string, value: string) => void;
+};
+
+type SessionManagerAuthParams =
```

Comment:
here every field is interconnected. You can't have default auth token when token login is disabled.
Static typing is not enough to prevent invalid states. We either use classes, of chage the way we model this

Reply:
Having a good runtime validation (maybe yup?) would be great, however here I just strive to have a good Typescript types.
I reworked this param type since it didn't make sense to have a case with all auth types disabled. If you think I am still mishandling the possible auth cases let me know.